tests/test_nhg.py

```diff
@@ -384,20 +799,891 @@ def asic_route_nhg_fvs(k):
+        rt_ps.set("2.2.2.0/24", fvs)
+
+        # check if route was propagated to ASIC DB
+
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 1)
+
+        k = self.get_route_id('2.2.2.0/24', dvs)
+
+        # assert the route points to next hop group
+        fvs = asic_db.get_entry(self.ASIC_RT_STR, k)
+        assert fvs["SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID"] == nhgid
+
+        # Bring group's link down
+        self.flap_intf(0, 'down', dvs)
+
+        # Check that the group still has the same ID and everything works
+        assert nhgid == self.get_nhg_id('group1', dvs)
+
+        # Bring group's link back up
+        self.flap_intf(0, 'up', dvs)
+
+        # Check that the group still has the same ID and everything works
+        assert nhgid == self.get_nhg_id('group1', dvs)
+
+        # Bring an interface down
+        self.flap_intf(1, 'down', dvs)
+
+        # Create group2 pointing to the NH which's link is down
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.3'), ("ifname", "Ethernet4")])
+        nhg_ps.set("group2", fvs)
+
+        # The group should be created.  To test this, add a route pointing to
+        # it.  If the group exists, the route will be created as well.
+        fvs = swsscommon.FieldValuePairs([("nexthop_group", "group2")])
+        rt_ps.set('2.2.4.0/24', fvs)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 2)
+
+        # Delete the route and the group
+        rt_ps._del('2.2.4.0/24')
+        nhg_ps._del('group2')
+
+        # Bring the interface back up
+        self.flap_intf(1, 'up', dvs)
+
+        # Create group2 pointing to a NH that does not exist
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.61'), ("ifname", "Ethernet120")])
+        nhg_ps.set("group2", fvs)
+
+        # The group should fail to be created.  To test this, we add a route
+        # pointing to it.  The route should not be created.
+        fvs = swsscommon.FieldValuePairs([("nexthop_group", "group2")])
+        rt_ps.set('2.2.4.0/24', fvs)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 1)
+
+        # Configure the NH's interface
+        self.config_intf(30, dvs)
+
+        # A couple of more routes will be added to ASIC_DB
+        asic_routes_count += 2
+
+        # The group should be created, so the route should be added
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 2)
+
+        # Delete the route and the group
+        rt_ps._del('2.2.4.0/24')
+        nhg_ps._del('group2')
+
+        # Update the group to a multiple NH group - should fail as the group
+        # is referenced
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.1,10.0.0.3'),
+                                          ("ifname", "Ethernet0,Ethernet4")])
+        nhg_ps.set("group1", fvs)
+        time.sleep(1)
+        assert len(asic_db.get_keys(self.ASIC_NHG_STR)) == asic_nhgs_count
+
```

Comment:
check SAI NHOP ID hasn't changed.

Reply:
Done.

Reply:
There are lot of tests in https://github.com/Azure/sonic-mgmt/blob/master/tests/

An example:
https://github.com/Azure/sonic-mgmt/blob/master/tests/qos/test_qos_sai.py